src/umpire/strategy/ResourceAwarePool.cpp

```diff
@@ -0,0 +1,549 @@
+
+  deallocate_resource(r, ptr, size);
+}
+
+void ResourceAwarePool::do_deallocate(Chunk* chunk, void* ptr) noexcept
+{
+  UMPIRE_POISON_MEMORY_REGION(m_allocator, ptr, chunk->size);
+  UMPIRE_USE_VAR(ptr);
+  chunk->free = true;
+
+  // Removing chunk from pending
+  for (auto it = m_pending_map.begin(); it != m_pending_map.end();) {
+    auto my_chunk = (*it);
+    if (my_chunk == chunk) {
+      m_pending_map.erase(it);
+    } else {
+      it++;
+    }
+  }
+
+  UMPIRE_LOG(Debug, "In the do_deallocate function. Deallocating data held by " << chunk);
+
+  if (chunk->prev && chunk->prev->free == true) {
+    auto prev = chunk->prev;
+
+    UMPIRE_LOG(Debug, "Removing chunk" << prev << " from free map");
+    m_free_map.erase(prev->size_map_it);
+
+    prev->size += chunk->size;
+    prev->next = chunk->next;
+
+    prev->m_event = chunk->m_event;
+    prev->m_resource = chunk->m_resource;
+
+    if (prev->next)
+      prev->next->prev = prev;
+
+    UMPIRE_LOG(Debug, "Merging with prev" << prev << " and " << chunk);
+    UMPIRE_LOG(Debug, "New size: " << prev->size);
+
+    chunk->~Chunk(); // manually call destructor
+    m_chunk_pool.deallocate(chunk);
+    chunk = prev;
+  }
+
+  if (chunk->next && chunk->next->free == true) {
+    auto next = chunk->next;
+
+    UMPIRE_LOG(Debug, "Removing chunk" << next << " from free map");
+    m_free_map.erase(next->size_map_it);
+
+    chunk->size += next->size;
+    chunk->next = next->next;
+
+    chunk->m_event = next->m_event;
+    chunk->m_resource = next->m_resource;
+
+    if (chunk->next)
+      chunk->next->prev = chunk;
+
+    UMPIRE_LOG(Debug, "Merging with next" << chunk << " and " << next);
+    UMPIRE_LOG(Debug, "New size: " << chunk->size);
+
+    next->~Chunk(); // manually call destructor
+    m_chunk_pool.deallocate(next);
+  }
+
+  UMPIRE_LOG(Debug, "Inserting chunk " << chunk << " with size " << chunk->size);
+
+  if (chunk->size == chunk->chunk_size) {
+    m_releasable_blocks++;
+    m_releasable_bytes += chunk->chunk_size;
+  }
+
+  chunk->size_map_it = m_free_map.insert(std::make_pair(chunk->size, chunk));
+}
+
+void ResourceAwarePool::deallocate_resource(camp::resources::Resource r, void* ptr, std::size_t UMPIRE_UNUSED_ARG(size))
+{
+  UMPIRE_LOG(Debug, "(ptr=" << ptr << ")");
+  UMPIRE_LOG(Debug, "(Resource=" << camp::resources::to_string(r) << ")");
+  auto chunk = (*m_used_map.find(ptr)).second;
+
+  if (chunk == nullptr) {
+    UMPIRE_ERROR(runtime_error, fmt::format("The chunk can't be found! Called deallocate with ptr: {}", ptr));
+  }
+
+  auto my_r = getResource(ptr);
+  if (my_r != r) {
+    UMPIRE_ERROR(
+        runtime_error,
+        fmt::format("Called deallocate with different resource than what is returned by getResource. Called with {},",
+                    "but getResource returned: {}", camp::resources::to_string(r), camp::resources::to_string(my_r)));
+  }
+
+  // Chunk is now pending, add to list
+  m_pending_map.push_back(chunk);
```

Comment:
```
    280   // Call deallocate logic only for a non-pending chunk
    281   if (chunk->m_event.check()) {
    282     do_deallocate(chunk, ptr);
    283   } else {
    284     // Chunk is now pending, add to list
    285     m_pending_map.push_back(chunk);
    286   }
 ```
 I have that instead - so after I create an event, if it immediately comes back as having completed (as with the Host resource), then I call do_deallocate, otherwise I add it to pending.